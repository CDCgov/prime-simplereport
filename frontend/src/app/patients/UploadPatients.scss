.usa-file-input,
.usa-file-input__target {
  display: flex !important;
  flex-direction: column;
}

.usa-file-input__preview {
  flex: auto;
}

.thick-bottom-border {
  border-bottom-width: 4px !important;
  width: 300px;
}

.maxw-710 {
  max-width: 44.375rem;
}

.maxw-560 {
  max-width: 35rem;
}

.prime-radio--horizontal__container {
  margin-right: 0.75rem !important;
}

.usa-file-input__target {
  margin-top: 0 !important;
}

.line-height-225 {
  line-height: 2.25rem !important;
}
