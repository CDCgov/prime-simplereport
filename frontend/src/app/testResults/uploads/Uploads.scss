.sr-test-results-uploads {
  .usa-process-list__heading {
    font-size: 1.17rem;
  }

  mark {
    background-color: #f0f0f0;
  }
}
