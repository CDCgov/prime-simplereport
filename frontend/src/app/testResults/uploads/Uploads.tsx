import React, { useState } from "react";
import { useSelector } from "react-redux";
import { Button, FormGroup, FileInput } from "@trussworks/react-uswds";

import { showError } from "../../utils/srToast";
import { FeedbackMessage } from "../../../generated/graphql";
import { useDocumentTitle } from "../../utils/hooks";
import { LinkWithQuery } from "../../commonComponents/LinkWithQuery";
import { FileUploadService } from "../../../fileUploadService/FileUploadService";
import "../HeaderSizeFix.scss";
import { getAppInsights } from "../../TelemetryService";
import { RootState } from "../../store";

const PAYLOAD_MAX_BYTES = 50 * 1000 * 1000;
const REPORT_MAX_ITEMS = 10000;
const REPORT_MAX_ITEM_COLUMNS = 2000;

const Uploads = () => {
  useDocumentTitle("Upload spreadsheet");

  const appInsights = getAppInsights();
  const orgName = useSelector<RootState, string>(
    (state) => state.organization?.name
  );
  const user = useSelector<RootState, User>((state) => state.user);

  const [fileInputResetValue, setFileInputResetValue] = useState(0);
  const [buttonIsDisabled, setButtonIsDisabled] = useState(true);
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [file, setFile] = useState<File>();

  const [reportId, setReportId] = useState<string | null>(null);

  const [errors, setErrors] = useState<
    Array<FeedbackMessage | undefined | null>
  >([]);
  const [errorMessageText, setErrorMessageText] = useState<string | null>(null);

  const handleFileChange = async (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    setButtonIsDisabled(true);
    setFile(undefined);

    try {
      if (!event?.currentTarget?.files?.length) {
        return; //no files
      }
      const currentFile = event.currentTarget.files.item(0);
      if (!currentFile) {
        return;
      }

      if (currentFile.size > PAYLOAD_MAX_BYTES) {
        const maxKBytes = (PAYLOAD_MAX_BYTES / 1024).toLocaleString("en-US", {
          maximumFractionDigits: 2,
          minimumFractionDigits: 2,
        });
        showError(
          `The file '${currentFile.name}' is too large.  The maximum file size is ${maxKBytes}k`,
          "Invalid file"
        );
        return;
      }

      const fileText = await currentFile.text();
      const lineCount = (fileText.match(/\n/g) || []).length + 1;
      if (lineCount > REPORT_MAX_ITEMS) {
        showError(
          `The file '${currentFile.name}' has too many rows. The maximum number of rows is ${REPORT_MAX_ITEMS}.`,
          "Invalid file"
        );
        return;
      }

      if (lineCount <= 1) {
        showError(
          `The file '${currentFile.name}' doesn't contain any valid data. File should have a header line and at least one line of data.`,
          "Invalid file"
        );
        return;
      }

      // get the first line and examine it
      const firstLine = (fileText.match(/^(.*)\n/) || [""])[0];
      // ideally, the columns would be comma seperated, but they may be tabs, because the first
      // line is a header, we don't have to worry about escaped delimiters in strings (e.g. ,"Smith, John",)
      const columnCount =
        (firstLine.match(/,/g) || []).length ||
        (firstLine.match(/\t/g) || []).length;

      if (columnCount > REPORT_MAX_ITEM_COLUMNS) {
        showError(
          `The file '${currentFile.name}' has too many columns. The maximum number of allowed columns is ${REPORT_MAX_ITEM_COLUMNS}.`,
          "Invalid file"
        );
        return;
      }
      setFile(currentFile);
      setButtonIsDisabled(false);
    } catch (err: any) {
      showError(err.toString(), "An unexpected error happened");
    }
  };

  const handleSubmit = async (event: React.FormEvent<HTMLButtonElement>) => {
    event.preventDefault();

    setIsSubmitting(true);
    setButtonIsDisabled(true);
    setReportId(null);
    setErrorMessageText(null);
    setErrors([]);

    if (!file || file.size === 0) {
      setErrorMessageText(
        "Please resolve the errors below and upload your edited file. Your file has not been accepted."
      );
      const errorMessage = {} as FeedbackMessage;
      errorMessage.message = "Invalid File";
      setErrors([errorMessage]);
      return;
    }

    FileUploadService.uploadResults(file).then(async (res) => {
      setIsSubmitting(false);
      setFileInputResetValue(fileInputResetValue + 1);
      setFile(undefined);

      if (res.status !== 200) {
        setErrorMessageText(
          "There was a server error. Your file has not been accepted."
        );
        appInsights?.trackEvent({
          name: "Spreadsheet upload server error",
          properties: {
            org: orgName,
            user: user?.email,
          },
        });
      } else {
        const response = await res.json();

        if (response?.reportId) {
          setReportId(response?.reportId);
          appInsights?.trackEvent({
            name: "Spreadsheet upload success",
            properties: {
              "report ID": response.reportId,
              org: orgName,
              user: user?.email,
            },
          });
        }

        if (response?.errors?.length) {
          setErrorMessageText(
            "Please resolve the errors below and upload your edited file. Your file has not been accepted."
          );
          setErrors(response.errors);
          appInsights?.trackEvent({
            name: "Spreadsheet upload validation failure",
            properties: {
              errors: response.errors,
              org: orgName,
              user: user?.email,
            },
          });
        }
      }
    });
  };

  return (
    <div className="grid-row header-size-fix">
      <div className="prime-container card-container">
        <div className="usa-card__header">
          <h1>Upload your results</h1>
        </div>
        <div className="usa-alert usa-alert--info margin-left-105em margin-right-105em maxw-tablet-lg">
          <div className="usa-alert__body">
            <h3 className="usa-alert__heading">
              What is the bulk results uploader?
            </h3>
            <p className="usa-alert__text">
              <em>
                This feature is in beta. That means it’s new, and we’ll continue
                to update and improve it for our users. Though the feature is in
                beta, it will route all the results you submit to the
                appropriate public health department(s). Results you upload here
                will not appear individually in the "Test results" page.
              </em>{" "}
              <br />
              <br /> The results uploader allows you to report test results in
              bulk using a CSV file. While we expect most SimpleReport users
              will continue to report through the regular process, this feature
              can serve labs and others with an information system that exports
              spreadsheets, such as an EMR.{" "}
              <LinkWithQuery to="/results/upload/submit/guide">
                <strong>Learn more about how it works</strong>
              </LinkWithQuery>
              .
            </p>
          </div>
        </div>
        <div className="usa-card__body padding-y-2 maxw-prose">
          <p>
            Report results in bulk using a comma-separated values (CSV)
            spreadsheet. To upload your spreadsheet:
          </p>
          <ol className="usa-list">
            <li>
              <a
                href="/assets/resources/test_results_example_10-3-2022.csv"
                onClick={() => {
                  appInsights?.trackEvent({
                    name: "Download spreadsheet template",
                  });
                }}
              >
                Download the spreadsheet template
              </a>{" "}
            </li>
            <li>
              Format and add your data following the template and upload guide
            </li>
            <li>Save your spreadsheet as a CSV</li>
            <li>Upload your CSV here</li>
          </ol>
          <p>
            For more information about preparing your spreadsheet, visit our{" "}
            <LinkWithQuery to="/results/upload/submit/guide">
              <strong>spreadsheet upload guide</strong>
            </LinkWithQuery>
            .
          </p>
          {reportId && (
            <div>
              <div className="usa-alert usa-alert--success">
                <div className="usa-alert__body">
                  <h3 className="usa-alert__heading">Success: File Accepted</h3>
                  <p className="usa-alert__text">
                    Your file has been successfully transmitted to the
                    department of health
                  </p>
                </div>
              </div>
              <div>
                <p className="text-normal text-base margin-bottom-0">
                  Confirmation Code
                </p>
                <p className="margin-top-05">{reportId}</p>
              </div>
            </div>
          )}
          {errorMessageText && (
            <div>
              <div className="usa-alert usa-alert--error" role="alert">
                <div className="usa-alert__body">
                  <h3 className="usa-alert__heading">
                    Error: File not accepted
                  </h3>
                  <p className="usa-alert__text">{errorMessageText}</p>
                </div>
              </div>
              {errors?.length > 0 && (
                <table className="usa-table usa-table--borderless">
                  <thead>
                    <tr>
                      <th>Requested Edit</th>
                      <th>Areas Containing the Requested Edit</th>
                    </tr>
                  </thead>
                  <tbody>
                    {errors.map((e, i) => {
                      return (
                        <tr key={"error_" + i}>
                          <td>{e?.["message"]} </td>
                          <td>Row(s): {e?.["indices"]}</td>
                        </tr>
                      );
                    })}
                  </tbody>
                </table>
              )}
            </div>
          )}
          <FormGroup className="margin-bottom-3">
            <FileInput
              key={fileInputResetValue}
              id="upload-csv-input"
              name="upload-csv-input"
              aria-label="Choose CSV file"
              accept="text/csv, .csv"
              onChange={(e) => handleFileChange(e)}
              required
            />
          </FormGroup>
          <Button
            type="submit"
            onClick={(e) => handleSubmit(e)}
            disabled={buttonIsDisabled || file?.name?.length === 0}
          >
            {isSubmitting && (
              <span>
                <span>Processing file...</span>
              </span>
            )}
            {!isSubmitting && <span>Upload your CSV</span>}
          </Button>
        </div>
      </div>
    </div>
  );
};

export default Uploads;
