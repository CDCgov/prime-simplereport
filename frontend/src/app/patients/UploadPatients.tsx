import React, { useState } from "react";
import { FileInput, FormGroup } from "@trussworks/react-uswds";
import { useSelector } from "react-redux";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faXmark } from "@fortawesome/free-solid-svg-icons";
import { useLocation } from "react-router-dom";

import { useDocumentTitle } from "../utils/hooks";
import Button from "../commonComponents/Button/Button";
import RadioGroup from "../commonComponents/RadioGroup";
import Dropdown from "../commonComponents/Dropdown";
import { Facility, FeedbackMessage } from "../../generated/graphql";
import { showError } from "../utils/srToast";
import { FileUploadService } from "../../fileUploadService/FileUploadService";
import iconLoader from "../../img/loader.svg";
import { getFacilityIdFromUrl } from "../utils/url";

import { AddPatientHeader } from "./Components/AddPatientsHeader";

import "./UploadPatients.scss";

const UploadPatients = () => {
  useDocumentTitle("Add Patient");
  const [facilityAmount, setFacilityAmount] = useState<string>();
  const [buttonIsDisabled, setButtonIsDisabled] = useState(true);
  const [selectedFacility, setSelectedFacility] = useState<Facility>();
  const [file, setFile] = useState<File>();
  const [errors, setErrors] = useState<
    Array<FeedbackMessage | undefined | null>
  >([]);
  const [errorMessageText, setErrorMessageText] = useState("");
  const [status, setStatus] = useState<
    "submitting" | "complete" | "success" | "fail" | ""
  >("");

  const facilities = useSelector(
    (state: any) => (state?.facilities as Facility[]) || []
  );
  const activeFacilityId = getFacilityIdFromUrl(useLocation());
  const facility = selectedFacility ||
    facilities.find((f) => f.id === activeFacilityId) ||
    facilities[0] || { id: "", name: "" };

  const onFacilitySelect = (e: React.ChangeEvent<HTMLSelectElement>) => {
    const selected = facilities.find((f) => f.id === e.target.value);
    if (selected) {
      setSelectedFacility(selected);
    }
  };

  const handleFileChange = async (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    try {
      if (!event?.currentTarget?.files?.length) {
        return; //no files
      }
      const currentFile = event.currentTarget.files.item(0);
      if (!currentFile) {
        return;
      }
      setFile(currentFile);
      setButtonIsDisabled(false);
      setStatus("");
    } catch (err: any) {
      showError(`An unexpected error happened: '${err.toString()}'`);
    }
  };

  const handleSubmit = async (event: React.FormEvent<HTMLButtonElement>) => {
    event.preventDefault();

    setStatus("submitting");
    setButtonIsDisabled(true);
    setErrors([]);
    setErrorMessageText("");

    if (!file || file.size === 0) {
      setStatus("fail");
      setButtonIsDisabled(false);
      setErrorMessageText("Invalid file");
      return;
    }
    const facilityId = facilityAmount === "oneFacility" ? facility.id : "";
    FileUploadService.uploadPatients(file, facilityId).then(async (res) => {
      setStatus("complete");
      setFile(undefined);
      setButtonIsDisabled(true);

      if (res.status !== 200) {
        setStatus("fail");
        setErrorMessageText(
          "There was a server error. Your file has not been accepted."
        );
      } else {
        const response = await res?.json();

        if (response.status === "FAILURE") {
          setStatus("fail");
          if (response?.errors?.length) {
            setErrorMessageText(
              "Please resolve the errors below and upload your edited file."
            );
            setErrors(response.errors);
          } else {
            setErrorMessageText(
              "There was a server error. Your file has not been accepted."
            );
          }
        } else {
          setStatus("success");
        }
      }
    });
  };

  return (
    <div className={"prime-edit-patient prime-home flex-1"}>
      <div className={"grid-container margin-bottom-4 maxw-desktop-lg"}>
        <div className="patient__header padding-bottom-0">
          <AddPatientHeader />
        </div>
        <div className={"card padding-top-5 padding-bottom-7"}>
          <ol className={"prime-ul margin-0"}>
            <li>
              <span className={"font-sans-lg text-light line-height-225"}>
                1. Setup your spreadsheet
              </span>
              <div className={"margin-left-3"}>
                <div
                  className={
                    "usa-summary-box margin-top-3 margin-bottom-3 margin-right-3 maxw-tablet-lg"
                  }
                >
                  <span className="usa-alert__heading text-bold">
                    Follow formatting guidelines
                  </span>
                  <p className="usa-alert__text margin-top-0">
                    To upload patients in bulk, spreadsheets need to match
                    SimpleReport data requirements. Follow the Patient bulk
                    upload guide closely to format your CSV before uploading it
                    below.
                  </p>
                </div>
                <Button className={"margin-right-105"}>
                  View patient bulk upload guide
                </Button>
                <Button variant={"outline"}>
                  Download spreadsheet template
                </Button>
              </div>
            </li>

            <li>
              <div
                className={
                  "margin-top-7 margin-bottom-1 font-sans-lg text-light line-height-225"
                }
              >
                2. Would you like to import these patients to one facility OR
                all facilities?
              </div>
              <div className={"margin-left-3"}>
                <div className={"maxw-710 line-height-sans-4"}>
                  If you plan to test patients at more than one facility, we
                  recommend adding them to all facilities. You can't select
                  multiple facilities individually.
                </div>
                <RadioGroup
                  wrapperClassName="margin-top-2"
                  inputClassName={"usa-radio__input--tile"}
                  name="facilitySector"
                  legend="Select facility"
                  legendSrOnly
                  buttons={[
                    {
                      value: "oneFacility",
                      label: "One facility",
                    },
                    {
                      value: "allFacility",
                      label: "All facilities",
                    },
                  ]}
                  selectedRadio={facilityAmount}
                  onChange={setFacilityAmount}
                  variant="horizontal"
                />
                {facilityAmount === "oneFacility" && (
                  <div className={"margin-top-205"}>
                    <div>Which facility?</div>
                    <Dropdown
                      aria-label={"Select facility"}
                      selectedValue={facility.id}
                      onChange={onFacilitySelect}
                      className={"grid-col-4"}
                      options={facilities.map(({ name, id }) => ({
                        label: name,
                        value: id,
                      }))}
                    />
                  </div>
                )}
              </div>
            </li>
            <li>
              <div
                className={
                  "margin-top-7 margin-bottom-1 font-sans-lg text-light line-height-225"
                }
              >
                3. Upload your spreadsheet
                {facilityAmount === "oneFacility" && " for " + facility.name}.
              </div>
              <div className={"margin-left-3"}>
                <div className={"maxw-710 line-height-sans-4"}>
                  The spreadsheet may take 10 or more minutes to upload. You do
                  not need to stay on this page. We'll email you if you need to
                  fix any errors or when the upload is complete.
                </div>
                <div>
                  {status === "success" && (
                    <div className="usa-alert usa-alert--success maxw-560 margin-top-3 outline-0">
                      <div className="usa-alert__body">
                        <span className="usa-alert__heading text-bold">
                          Success: File Accepted
                        </span>
                        <button
                          className="Toastify__close-button Toastify__close-button--default position-absolute top-0 right-0"
                          type="button"
                          aria-label="close"
                          onClick={() => setStatus("")}
                        >
                          <FontAwesomeIcon icon={faXmark} />
                        </button>
                        <p className="usa-alert__text">
                          Patients in your file have been successfully uploaded.
                        </p>
                      </div>
                    </div>
                  )}
                  {status === "fail" && (
                    <div className={"margin-top-3"}>
                      {errorMessageText && (
                        <div className="usa-alert usa-alert--error maxw-560">
                          <div className="usa-alert__body">
                            <span className="usa-alert__heading text-bold">
                              Error: File not accepted
                            </span>
                            <button
                              className="Toastify__close-button Toastify__close-button--default position-absolute top-0 right-0"
                              type="button"
                              aria-label="close"
                              onClick={() => {
                                setErrorMessageText("");
                                setErrors([]);
                              }}
                            >
                              <FontAwesomeIcon icon={faXmark} />
                            </button>
                            <p className="usa-alert__text">
                              {errorMessageText}
                            </p>
                          </div>
                        </div>
                      )}
                      {errors.length > 0 && (
                        <table className="usa-table usa-table--borderless">
                          <thead>
                            <tr>
                              <th
                                className={"thick-bottom-border padding-left-0"}
                              >
                                Edits needed
                              </th>
                              <th
                                className={"thick-bottom-border padding-left-0"}
                              >
                                Areas requiring edits
                              </th>
                            </tr>
                          </thead>
                          <tbody>
                            {errors.map((e, i) => {
                              return (
                                <tr key={"error_" + i}>
                                  <td className={"border-bottom-0"}>
                                    {e?.["message"]}{" "}
                                  </td>
                                  <td className={"border-bottom-0"}>
                                    {e?.["indices"] &&
                                      "Row(s): " + e?.["indices"]}
                                  </td>
                                </tr>
                              );
                            })}
                          </tbody>
                        </table>
                      )}
                    </div>
                  )}
                  <FormGroup className="margin-bottom-3">
                    {status === "submitting" ? (
                      <div className={"usa-file-input"}>
                        <div className={"usa-file-input__target"}>
                          <div className={"margin-top-1"}>
                            <span className="usa-file-input__drag-text font-sans-xs text-bold">
                              Uploading patient information...
                            </span>
                            <div className={"margin-top-1"}>
                              <img
                                src={iconLoader}
                                alt="submitting"
                                className={"square-5"}
                              />
                            </div>
                          </div>
                        </div>
                      </div>
                    ) : (
                      <FileInput
                        id="upload-csv-input"
                        name="upload-csv-input"
                        aria-label="Choose CSV file"
                        accept="text/csv, .csv"
                        onChange={handleFileChange}
                        required
                      />
                    )}
                  </FormGroup>
                </div>
                <div>
                  <Button
                    disabled={buttonIsDisabled || facilityAmount === undefined}
                    onClick={handleSubmit}
                  >
                    Upload CSV file
                  </Button>
                </div>
              </div>
            </li>
          </ol>
        </div>
      </div>
    </div>
  );
};

export default UploadPatients;
