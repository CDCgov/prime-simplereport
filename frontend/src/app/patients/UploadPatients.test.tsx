import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import createMockStore from "redux-mock-store";
import { MemoryRouter } from "react-router-dom";
import React from "react";
import userEvent from "@testing-library/user-event";

import { file } from "../testResults/uploads/Uploads.test";
import { FileUploadService } from "../../fileUploadService/FileUploadService";

import UploadPatients from "./UploadPatients";

const mockStore = createMockStore([]);
const store = mockStore({
  facilities: [
    { id: "1", name: "Lincoln Middle School" },
    { id: "2", name: "Rosa Parks High School" },
    { id: "3", name: "Empty School" },
  ],
});
const renderUploadPatients = () =>
  render(
    <Provider store={store}>
      <MemoryRouter>
        <UploadPatients />
      </MemoryRouter>
    </Provider>
  );
const uploadPatientsSpy = (response: Response) =>
  jest.spyOn(FileUploadService, "uploadPatients").mockImplementation(() => {
    return Promise.resolve(response);
  });
const errorResponseBody = {
  status: "FAILURE",
  errors: [{ indices: [0], message: "bad zipcode" }],
};

const successResponseBody = {
  status: "SUCCESS",
  errors: null,
};

function userEventUpload(uploadFile: File, facility: string) {
  const input = screen.getByTestId("file-input-input");
  userEvent.upload(input, uploadFile);
  userEvent.click(screen.getByText(facility));
  userEvent.click(screen.getByText("Upload CSV file"));
}

const submitCSVFile = (
  mockResponse: Response,
  uploadFile = file("someText"),
  facility = "All facilities"
) => {
  let spy = uploadPatientsSpy(mockResponse);
  userEventUpload(uploadFile, facility);
  return spy;
};

describe("Upload Patient", () => {
  beforeEach(() => {});
  afterEach(() => {
    jest.clearAllMocks();
  });
  it("displays the upload patients page correctly", () => {
    const { container } = renderUploadPatients();
    expect(container).toMatchSnapshot();
  });
  describe("facility selector", () => {
    it("should add facility name to label when one facility is selected", async () => {
      renderUploadPatients();
      expect(
        await screen.findByText("3. Upload your spreadsheet.")
      ).toBeInTheDocument();

      userEvent.click(screen.getByText("One facility"));

      expect(await screen.findByText("Which facility?")).toBeInTheDocument();
      expect(
        await screen.findByText(
          "3. Upload your spreadsheet for Lincoln Middle School."
        )
      );
    });
    it("should remove facility name when all facilities is selected", async () => {
      renderUploadPatients();
      expect(
        await screen.findByText("3. Upload your spreadsheet.")
      ).toBeInTheDocument();
      userEvent.click(screen.getByText("One facility"));

      userEvent.click(screen.getByText("All facilities"));
      expect(
        await screen.findByText("3. Upload your spreadsheet.")
      ).toBeInTheDocument();
    });
  });
  describe("submit button should be disabled", () => {
    it("Until facility and file have values", async () => {
      renderUploadPatients();
      expect(await screen.findByText("Upload CSV file")).toBeDisabled();

      userEvent.click(screen.getByText("One facility"));
      expect(await screen.findByText("Upload CSV file")).toBeDisabled();
      const uploadFile = file("someText");
      const input = screen.getByTestId("file-input-input");
      userEvent.upload(input, uploadFile);

      expect(await screen.findByText("Upload CSV file")).toBeEnabled();
    });
    it("Until file and facility have values", async () => {
      renderUploadPatients();
      expect(await screen.findByText("Upload CSV file")).toBeDisabled();

      const uploadFile = file("someText");
      const input = screen.getByTestId("file-input-input");
      userEvent.upload(input, uploadFile);
      expect(await screen.findByText("Upload CSV file")).toBeDisabled();
      userEvent.click(screen.getByText("One facility"));

      expect(await screen.findByText("Upload CSV file")).toBeEnabled();
    });
  });
  it("should upload to single facility", async () => {
    renderUploadPatients();
    let mockResponse = new Response(JSON.stringify(successResponseBody), {
      status: 200,
    });
    const uploadFile = file("someText");

    const uploadSpy = submitCSVFile(mockResponse, uploadFile, "One facility");

    expect(uploadSpy).toHaveBeenCalledWith(uploadFile, "1");
    expect(
      await screen.findByText("Success: File Accepted")
    ).toBeInTheDocument();
  });
  it("should show success message if upload is successful", async () => {
    renderUploadPatients();
    let mockResponse = new Response(JSON.stringify(successResponseBody), {
      status: 200,
    });
    const uploadFile = file("someText");

    const uploadSpy = submitCSVFile(mockResponse, uploadFile);

    expect(uploadSpy).toHaveBeenCalledWith(uploadFile, "");
    expect(
      await screen.findByText("Success: File Accepted")
    ).toBeInTheDocument();
  });
  it("should show error message and list errors if error occurs", async () => {
    renderUploadPatients();
    let mockResponse = new Response(JSON.stringify(errorResponseBody), {
      status: 200,
    });

    submitCSVFile(mockResponse);

    expect(
      await screen.findByText("Error: File not accepted")
    ).toBeInTheDocument();
    expect(
      await screen.findByText(
        "Please resolve the errors below and upload your edited file."
      )
    ).toBeInTheDocument();
    expect(await screen.findByText("bad zipcode")).toBeInTheDocument();
    expect(await screen.findByText("Row(s): 0")).toBeInTheDocument();
  });
  it("should show error message if 500 is returned", async () => {
    renderUploadPatients();
    let mockResponse = new Response(null, {
      status: 500,
    });
    const uploadFile = file("someText");

    const uploadSpy = submitCSVFile(mockResponse, uploadFile);

    expect(uploadSpy).toHaveBeenCalledWith(uploadFile, "");
    expect(
      await screen.findByText("Error: File not accepted")
    ).toBeInTheDocument();
  });
  it("should close success message when close is clicked", async () => {
    renderUploadPatients();
    let mockResponse = new Response(JSON.stringify(successResponseBody), {
      status: 200,
    });
    uploadPatientsSpy(mockResponse);
    submitCSVFile(mockResponse);
    expect(
      await screen.findByText("Success: File Accepted")
    ).toBeInTheDocument();

    userEvent.click(screen.getByLabelText("close"));

    expect(
      screen.queryByText("Success: File Accepted")
    ).not.toBeInTheDocument();
  });
  it("should remove error message and table when close is clicked", async () => {
    renderUploadPatients();
    let mockResponse = new Response(JSON.stringify(errorResponseBody), {
      status: 200,
    });

    submitCSVFile(mockResponse);
    expect(
      await screen.findByText("Error: File not accepted")
    ).toBeInTheDocument();
    expect(await screen.findByText("bad zipcode")).toBeInTheDocument();

    userEvent.click(screen.getByLabelText("close"));
    expect(
      screen.queryByText("Error: File not accepted")
    ).not.toBeInTheDocument();
    expect(screen.queryByText("bad zipcode")).not.toBeInTheDocument();
  });
  it("should show loading message when file is being processed", async () => {
    renderUploadPatients();
    jest
      .spyOn(FileUploadService, "uploadPatients")
      .mockImplementation(async () => {
        await new Promise((resolve) => setTimeout(resolve, 10000));
        return Promise.resolve(new Response());
      });

    userEventUpload(file("abc"), "All facilities");

    expect(
      await screen.findByText("Uploading patient information...")
    ).toBeInTheDocument();
  });
  it("should show error if empty file is provided", async () => {
    renderUploadPatients();

    userEventUpload(file(""), "All facilities");

    expect(screen.getByText("Invalid file")).toBeInTheDocument();
    expect(
      screen.queryByText("Uploading patient information...")
    ).not.toBeInTheDocument();
  });
  describe("handle file change", () => {
    it("should do nothing if no file was added", () => {
      renderUploadPatients();

      const input = screen.getByTestId("file-input-input");
      userEvent.upload(input, []);

      expect(screen.getByText("Drag file here or")).toBeInTheDocument();
    });
    it("shows the file when selected", () => {
      renderUploadPatients();

      const input = screen.getByTestId("file-input-input");
      userEvent.upload(input, file("someText"));

      expect(screen.getByText("Selected file")).toBeInTheDocument();
      expect(screen.getByText("values.csv")).toBeInTheDocument();
    });
  });
});
