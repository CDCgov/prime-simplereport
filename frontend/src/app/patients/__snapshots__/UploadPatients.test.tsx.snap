// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload Patient displays the upload patients page correctly 1`] = `
<div>
  <div
    class="prime-edit-patient prime-home flex-1"
  >
    <div
      class="grid-container margin-bottom-4 maxw-desktop-lg"
    >
      <div
        class="patient__header padding-bottom-0"
      >
        <div
          class="display-flex flex-justify"
        >
          <div>
            <div
              class="display-flex flex-align-center"
            >
              <svg
                aria-hidden="true"
                class="usa-icon text-base margin-left-neg-2px"
                focusable="false"
                role="img"
              >
                <use
                  xlink:href="sprite.svg#arrow_back"
                />
              </svg>
              <a
                class="margin-left-05"
                href="/patients"
              >
                People
              </a>
            </div>
            <div
              class="prime-edit-patient-heading margin-y-0"
            >
              <h1
                class="font-heading-lg margin-top-1 margin-bottom-0"
              >
                Add new 
                patient
              </h1>
            </div>
          </div>
        </div>
        <nav
          aria-label="Secondary navigation"
          class="prime-secondary-nav padding-top-2"
        >
          <ul
            class="usa-nav__secondary-links prime-nav"
          >
            <li
              class="usa-nav__secondary-item padding-left-0"
            >
              <a
                class=""
                href="/add-patient"
              >
                Add individual 
                patient
              </a>
            </li>
            <li
              class="usa-nav__secondary-item"
            >
              <a
                class=""
                href="/upload-patients"
              >
                Import from spreadsheet
              </a>
            </li>
          </ul>
        </nav>
      </div>
      <div
        class="card padding-top-5 padding-bottom-7"
      >
        <ol
          class="prime-ul margin-0"
        >
          <li>
            <span
              class="font-sans-lg text-light line-height-225"
            >
              1. Setup your spreadsheet
            </span>
            <div
              class="margin-left-3"
            >
              <div
                class="usa-summary-box margin-top-3 margin-bottom-3 margin-right-3 maxw-tablet-lg"
              >
                <span
                  class="usa-alert__heading text-bold"
                >
                  Follow formatting guidelines
                </span>
                <p
                  class="usa-alert__text margin-top-0"
                >
                  To upload patients in bulk, spreadsheets need to match SimpleReport data requirements. Follow the Patient bulk upload guide closely to format your CSV before uploading it below.
                </p>
              </div>
              <button
                class="usa-button margin-right-105"
                type="button"
              >
                View patient bulk upload guide
              </button>
              <button
                class="usa-button usa-button--outline"
                type="button"
              >
                Download spreadsheet template
              </button>
            </div>
          </li>
          <li>
            <div
              class="margin-top-7 margin-bottom-1 font-sans-lg text-light line-height-225"
            >
              2. Would you like to import these patients to one facility OR all facilities?
            </div>
            <div
              class="margin-left-3"
            >
              <div
                class="maxw-710 line-height-sans-4"
              >
                If you plan to test patients at more than one facility, we recommend adding them to all facilities. You can't select multiple facilities individually.
              </div>
              <div
                class="usa-form-group margin-top-2"
              >
                <fieldset
                  class="usa-fieldset prime-radios"
                  id="facilitySector"
                >
                  <legend
                    class="usa-legend usa-sr-only"
                  >
                    Select facility
                  </legend>
                  <div
                    class="prime-radio--horizontal"
                  >
                    <div
                      class="usa-radio prime-radio--horizontal__container"
                    >
                      <input
                        class="usa-radio__input usa-radio__input--tile"
                        data-required="false"
                        id="1-val-oneFacility"
                        name="facilitySector"
                        type="radio"
                        value="oneFacility"
                      />
                      <label
                        class="usa-radio__label"
                        for="1-val-oneFacility"
                      >
                        One facility
                      </label>
                    </div>
                    <div
                      class="usa-radio prime-radio--horizontal__container"
                    >
                      <input
                        class="usa-radio__input usa-radio__input--tile"
                        data-required="false"
                        id="1-val-allFacility"
                        name="facilitySector"
                        type="radio"
                        value="allFacility"
                      />
                      <label
                        class="usa-radio__label"
                        for="1-val-allFacility"
                      >
                        All facilities
                      </label>
                    </div>
                  </div>
                </fieldset>
              </div>
            </div>
          </li>
          <li>
            <div
              class="margin-top-7 margin-bottom-1 font-sans-lg text-light line-height-225"
            >
              3. Upload your spreadsheet
              .
            </div>
            <div
              class="margin-left-3"
            >
              <div
                class="maxw-710 line-height-sans-4"
              >
                The spreadsheet may take 10 or more minutes to upload. You do not need to stay on this page. We'll email you if you need to fix any errors or when the upload is complete.
              </div>
              <div>
                <div
                  class="usa-form-group margin-bottom-3"
                  data-testid="formGroup"
                >
                  <div
                    class="usa-file-input"
                    data-testid="file-input"
                  >
                    <div
                      class="usa-file-input__target"
                      data-testid="file-input-droptarget"
                    >
                      <div
                        aria-hidden="true"
                        class="usa-file-input__instructions"
                        data-testid="file-input-instructions"
                      >
                        <span
                          class="usa-file-input__drag-text"
                        >
                          Drag file here or 
                        </span>
                        <span
                          class="usa-file-input__choose"
                        >
                          choose from folder
                        </span>
                      </div>
                      <div
                        class="usa-file-input__box"
                        data-testid="file-input-box"
                      />
                      <input
                        accept="text/csv, .csv"
                        aria-label="Choose CSV file"
                        class="usa-file-input__input"
                        data-testid="file-input-input"
                        id="upload-csv-input"
                        name="upload-csv-input"
                        required=""
                        type="file"
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div>
                <button
                  aria-disabled="true"
                  class="usa-button usa-button-disabled"
                  disabled=""
                  type="button"
                >
                  Upload CSV file
                </button>
              </div>
            </div>
          </li>
        </ol>
      </div>
    </div>
  </div>
</div>
`;
